### PR TITLE
Add support of customizing RNG for crypto algorithms

### PR DESCRIPTION
**Pull request by Taowyoo**
## TL;DR

This PR picks changes in https://github.com/fortanix/rustls-mbedtls-provider/pull/64 to master.

## Background

Current two RNG we provided in this crate does not meet FIPS requirements.

So, this PR updates all crypto algorithms that need to use RNG to have a function pointer for creating RNG. In this way, user could customize the RNG they want to use for each algorithm.

## Changes

- Update code and add `fn with_rng_provider` to enable user to choose RNG for each crypto algorithm implementation.
- Make some more types `pub`, so user could easily customize current crypto algorithm implementations.
- Add unit tests.
- Fix latest warnings from nightly rust:
  - Remove unnecessary namespace path used in types.
  - Remove unnecessary `Deref` trait guard.

### CODE REVIEW

**Review by xinyufort — APPROVED** [review]

File: rustls-mbedcrypto-provider/src/agreement.rs [hunk 1]
```diff
@@ -8,6 +8,7 @@
 use mbedtls::pk::{EcGroupId, ECDSA_MAX_LEN};
 
 /// An ECDH key agreement algorithm.
+#[derive(Debug)]
```
Comment (arai-fortanix) [inline rustls-mbedcrypto-provider/src/agreement.rs:11]: You might want to think about whether you really want to derive Debug for this struct. While it doesn't contain any secrets, knowing these values might have value for an attacker.

Comment (arai-fortanix) [reply rustls-mbedcrypto-provider/src/agreement.rs:11]: I think Francisco should probably look at this change. I'm out of my depth both for some of the Rust constructions and for reviewing crypto code.

Comment (Taowyoo) [reply rustls-mbedcrypto-provider/src/agreement.rs:11]: Make sense. I can actually avoid this line, since it's not necssary to debug print them (There is no clear use case to print them).

Comment (Taowyoo) [reply rustls-mbedcrypto-provider/src/agreement.rs]: removed

Comment (zugzwang) [reply rustls-mbedcrypto-provider/src/agreement.rs]: In the same sense, is https://github.com/fortanix/rustls-mbedtls-provider/pull/65/files#diff-508887ea2da6ea0a0f89bc3212a612ca92878ca711ed5f8599604cdbc919a719R55 strictly necessary?

File: rustls-mbedcrypto-provider/src/lib.rs [hunk 2]
```diff
@@ -117,19 +118,27 @@ pub mod rng {
     #[cfg(not(any(target_env = "sgx", feature = "rdrand")))]
     use mbedtls::rng::{CtrDrbg, OsEntropy};
 
-    #[cfg(any(target_env = "sgx", feature = "rdrand"))]
-    use mbedtls::rng::Rdrand;
+    /// Type alias for the RNG supported by *mbedtls*
+    #[cfg(not(any(target_env = "sgx", feature = "rdrand")))]
+    pub type MbedRng = CtrDrbg;
 
     /// Get a RNG supported by *mbedtls*
```
[the remark this answers is not in the document]
Comment (Taowyoo) [reply rustls-mbedcrypto-provider/src/lib.rs:125]: Sure, will do!

File: rustls-mbedcrypto-provider/src/kx.rs [hunk 3]
```diff
@@ -28,34 +30,50 @@ use rustls::ffdhe_groups;
...
+pub struct KxGroup<T: RngCallback> {
     /// The IANA "TLS Supported Groups" name of the group
     name: NamedGroup,
 
-    /// The corresponding [`agreement::Algorithm`]
+    /// The corresponding agreement algorithm
     agreement_algorithm: &'static agreement::Algorithm,
+
+    /// Callback to produce RNGs when needed
+    rng_provider: fn() -> Option<T>,
```
[the remark this answers is not in the document]
Comment (zugzwang) [reply rustls-mbedcrypto-provider/src/kx.rs:45]: I understand that key generation (and potentially, group action) need RNG, but it is not an intrinsic part a Key Exchange group as concept. It's just an operation that uses the key exchange group and the RNG provider. I agree that a global RNG is not a better approach. Is there no other design?

File: rustls-mbedcrypto-provider/src/agreement.rs (hunk 1 above)
Comment (zugzwang) [reply rustls-mbedcrypto-provider/src/agreement.rs]: Got it. This is another reason to not include `rng_provider` in the trait. Debugging the supported Key Exchange group is harmless (a TLS server obviously advertises this), but not the RNG used by the implementation.

Comment (zugzwang) [reply rustls-mbedcrypto-provider/src/agreement.rs]: On the very least, we could not debug out the `rng_provider` field

Comment (Taowyoo) [reply rustls-mbedcrypto-provider/src/agreement.rs]: Got your idea. I will remove them.

Comment (Taowyoo) [reply rustls-mbedcrypto-provider/src/agreement.rs]: done

File: rustls-mbedcrypto-provider/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Taowyoo) [reply rustls-mbedcrypto-provider/src/lib.rs]: done

File: rustls-mbedcrypto-provider/src/kx.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (zugzwang) [reply rustls-mbedcrypto-provider/src/kx.rs:45]: If there's no way around it, I suggest defining a wrapper struct that has the RNG provider and the key exchange group.

[the remark this answers is not in the document]
Comment (Taowyoo) [reply rustls-mbedcrypto-provider/src/kx.rs]: updated

**Comment by Taowyoo** [comment]
Hi folks, I just refactor and create the wrapper type in a clearer way:

- Refactor and rename `KxGroup` to `KxGroupWrapper`.
- Rename `KeyExchange` to `KeyExchangeImpl`.
- Refactor and rename `DheKxGroup` to `DheKxGroupWrapper`.
- Rename `DheActiveKeyExchange` to `DheActiveKeyExchangeImpl`.
- Refactor and rename `MbedTlsPkSigningKey` to `MbedTlsPkSigningKeyWrapper`.
- Update unit tests accordingly.

I did not change all types that contains `rng_provider` because the name of those unchanged type is fine to contain `rng_provider` in it like `MbedTlsSigner`.

CC: @zugzwang @xinyufort

File: rustls-mbedcrypto-provider/src/fips_utils/mod.rs [hunk 4]
```diff
@@ -318,18 +322,18 @@ mod tests {
...
-        let g = Mpi::from_binary(dhe_group.group.g).unwrap();
-        let p = Mpi::from_binary(dhe_group.group.p).unwrap();
+    fn create_ffdhe_key_pair<T: RngCallback>(dhe_group: &DheKxGroupWrapper<T>) -> (Mpi, Mpi) {
+        let g = Mpi::from_binary(dhe_group.dhe_kx_group.group.g).unwrap();
+        let p = Mpi::from_binary(dhe_group.dhe_kx_group.group.p).unwrap();
         let mut rng = crate::rng::rng_new().unwrap();
-        let mut x_binary = vec![0; dhe_group.priv_key_len];
+        let mut x_binary = vec![0; dhe_group.dhe_kx_group.priv_key_len];
         rng.random(&mut x_binary).unwrap();
         print_vec("private", &x_binary);
```
Comment (zugzwang) [inline rustls-mbedcrypto-provider/src/fips_utils/mod.rs:331]: This is a test function but please remove these `print_vec` statements with secret values

File: rustls-mbedcrypto-provider/src/kx.rs [hunk 5]
```diff
@@ -28,110 +30,181 @@ use rustls::ffdhe_groups;
 use rustls::ffdhe_groups::FfdheGroup;
 use rustls::Error;
 use rustls::NamedGroup;
-/// A key-exchange group supported by *mbedtls*.
+/// A wrapper type representing implmentation of an EC key-exchange group
```
Comment (zugzwang) [inline rustls-mbedcrypto-provider/src/kx.rs:33]: implmentation -> implementation

Also this is not specific to EC, and this comment doesn't explain much.

File: rustls-mbedcrypto-provider/src/kx.rs [hunk 6]
```diff
@@ -28,110 +30,181 @@ use rustls::ffdhe_groups;
...
+/// supported by *mbedtls*.
 ///
-/// All possible instances of this type are provided by the library in
-/// the `ALL_KX_GROUPS` array.
+/// All possible instances of this type are provided by the library in the
+/// `ALL_KX_GROUPS` array.
+pub struct KxGroupWrapper<T: RngCallback> {
+    /// An EC key-exchange group
```
Comment (zugzwang) [inline rustls-mbedcrypto-provider/src/kx.rs:39]: Not necessarily EC

File: rustls-mbedcrypto-provider/src/kx.rs (hunk 5 above)
Comment (zugzwang) [reply rustls-mbedcrypto-provider/src/kx.rs:33]: Huh, looking at the code, this is indeed EC, so this is a naming issue. So, this struct should be called `EcdhKxGroupWrapper` or so. And then, `DheKxGroupWrapper` should be `FfdheKxGroupWrapper`.

Could you try unify both implementations? So that `KxGroup` is able to represent both ECDH and FFDH? So it can be called `KxGroup`

**Review by zugzwang — CHANGES_REQUESTED** [review]
`KxGroup` should represent either an ECDH or FFDH key exchange group.

Currently, `KxGroup` represents ECDH and `DheKxGroup` represents FFDH.

File: rustls-mbedcrypto-provider/src/kx.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (s-arash) [reply rustls-mbedcrypto-provider/src/kx.rs]: FWIW, I'm not in favor of this change. It just adds unnecessary complexity. The `KxGroup` types are implementations of KX group required for rustls integration, not some abstract notion of key exchange groups.

File: rustls-mbedcrypto-provider/src/kx.rs (hunk 5 above)
Comment (Taowyoo) [reply rustls-mbedcrypto-provider/src/kx.rs:33]: I will update them to use better names. 

Regarding the unification. I did not see the strong reason to unify them into one type especially why the algorithm and upstream trait implementation are different.

Note: name of `KxGroup` is first introduced when ffdhe is not added, so I think the problem here is we just need better naming for two types.

File: rustls-mbedcrypto-provider/src/fips_utils/mod.rs [hunk 7]
```diff
@@ -17,13 +17,14 @@ use mbedtls::{
...
     pk::{EcGroupId, Pk},
+    rng::RngCallback,
 };
 
 mod constants;
 
 use crate::{
     fips_utils::constants::{get_ffdhe_q, get_known_ec_key, get_known_ffdhe_key_pair},
-    kx::DheKxGroup,
+    kx::{FFdheKxGroupWrapper, FfdheKxGroup},
```
Comment (zugzwang) [inline rustls-mbedcrypto-provider/src/fips_utils/mod.rs:27]: using `Ffdhe` and `FFdhe`, choose one

**Review by zugzwang — APPROVED** [review]
Small nit; use either `Ffdhe` or `FFDhe`, not both.

File: rustls-mbedcrypto-provider/src/kx.rs [hunk 8]
```diff
@@ -28,110 +30,181 @@ use rustls::ffdhe_groups;
...
+    rng_provider: crate::rng::rng_new,
 };
 
 /// DHE group [FFDHE4096](https://www.rfc-editor.org/rfc/rfc7919.html#appendix-A.3)
 pub static FFDHE4096: &dyn SupportedKxGroup = FFDHE4096_KX_GROUP;
-pub(crate) static FFDHE4096_KX_GROUP: &DheKxGroup = &DheKxGroup {
-    named_group: NamedGroup::FFDHE4096,
-    group: ffdhe_groups::FFDHE4096,
-    priv_key_len: 48,
+/// DHE group [FFDHE3072](https://www.rfc-editor.org/rfc/rfc7919.html#appendix-A.2)
```
Comment (xinyufort) [inline rustls-mbedcrypto-provider/src/kx.rs:176]: Wrong doc comment?

**Review by xinyufort — APPROVED** [review]
just one nitpick, otherwise LGTM

**Review by xinyufort — APPROVED** [review]
